lgc/util/llpcGfxRegHelper.h

```diff
@@ -0,0 +1,145 @@
+ *
+ *  Copyright (c) 2020 Advanced Micro Devices, Inc. All Rights Reserved.
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+ /**
+  ***********************************************************************************************************************
+  * @file  llpcGfxRegHelper.h
+  * @brief LLPC header file: contains the definition of LLPC utility class GfxRegHelper.
+  ***********************************************************************************************************************
+  */
+
+#pragma once
+#include "llpcGfxRegHandler.h"
+#include "llpcTargetInfo.h"
+
+namespace lgc
+{
+
+using namespace llvm;
+
+// =====================================================================================================================
+// Struct maintains state for specific register
+// - pValue : Cache the value of the register which is being fetched already.
+// - isModified : Set to be true if the register is being settled with new value.
+struct BitsState
+{
+    Value* pValue = nullptr;
+    bool isModified = false;
+};
+
+// =====================================================================================================================
+// Helper class for handling graphics registers
+class GfxRegHelper : public GfxRegHandlerBase
+{
+protected:
+    GfxRegHelper(Builder* pBuilder, Value* pRegister);
+
+    virtual ~GfxRegHelper() = default;
+
+    // Get register
+    Value* GetRegCommon(uint32_t id);
+
+    // Set register
+    void SetRegCommon(uint32_t id, Value* pVal)
+    {
+        SetBits(m_bitsInfo[id], pVal);
+        // It is assumed the register is being modified
+        m_bitsState[id].isModified = 1;
+    }
+
+    // Get combined data from two seperate DWORDs
+    Value* GetRegCombine(uint32_t idLo, uint32_t idHi);
```

Comment:
What is the llvm type of the returned value? I imagine that it'd be a 2-dword-large integer type, or maybe a vector type?

Reply:
The return time is a dword, it doesn't support 2-dword for now.
```
Dword 0
{
    unused                : 30 
    width_low            :  2
}

Dword 1
{
    width_hi               :  12
    unused                :  20 
}
```
I had considered the naming of this function and there should be some restirctions to the low and high BitsInfo. But, there's no usage that really need a 2-dword output value.

Reply:
I think this is important and should be included in the comment. I agree that finding a good name is difficult here.

Reply:
Yes, i'll comment this information.